services/paymaster/src/api.rs

```diff
@@ -157,6 +172,231 @@ pub async fn validate_transaction(
     Ok(())
 }
 
+pub fn validate_transaction_against_variation(
+    transaction: &VersionedTransaction,
+    tx_variation: &TransactionVariation,
+    sponsor: &Pubkey,
+) -> Result<(), (StatusCode, String)> {
+    match tx_variation {
+        TransactionVariation::V1(variation) => validate_transaction_against_variation_v1(transaction, variation, sponsor),
+    }
+}
+
+// TODO: incorporate gas spend and rate limit checks
+pub fn validate_transaction_against_variation_v1(
+    transaction: &VersionedTransaction,
+    variation: &crate::constraint::TransactionVariationV1,
+    sponsor: &Pubkey,
+) -> Result<(), (StatusCode, String)> {
+    let instructions = transaction.message.instructions();
+    if instructions.len() != variation.instructions.len() {
+        return Err((
+            StatusCode::BAD_REQUEST,
+            format!(
+                "Transaction instruction count {} does not match expected count {} for variation {}",
+                instructions.len(),
+                variation.instructions.len(),
+                variation.name
+            ),
+        ));
+    }
+
+    for (instr, constraint) in instructions.iter().zip(&variation.instructions) {
+        let program_id = instr.program_id(transaction.message.static_account_keys());
+        if *program_id != constraint.program {
+            return Err((
+                StatusCode::BAD_REQUEST,
+                format!(
+                    "Transaction instruction program ID {} does not match expected ID {} for variation {}",
+                    program_id,
+                    constraint.program,
+                    variation.name
+                ),
+            ));
+        }
+
+        for (i, account_constraint) in constraint.accounts.iter().enumerate() {
+            // TODO: account for lookup tables
+            let account_index = instr.accounts.get(account_constraint.index as usize).ok_or_else(|| {
+                (
+                    StatusCode::BAD_REQUEST,
+                    format!(
+                        "Transaction instruction missing account at index {} for variation {}",
+                        i, variation.name
+                    ),
+                )
+            })?;
+            let account = transaction.message.static_account_keys()[*account_index as usize];
+            // TODO: do fix this...
+            let user_dummy = &Pubkey::new_unique();
+            let session_dummy = &Pubkey::new_unique();
+            check_account_constraint(account, account_constraint, user_dummy, session_dummy, sponsor)?;
+        }
+
+        for data_constraint in &constraint.data {
+            check_data_constraint(&instr.data, data_constraint)?;
+        }
+    }
+
+    Ok(())
+}
+
+pub fn check_account_constraint(
```

Comment:
Should we think about making this function able to represent either a limit to the number of accounts, or a filter accepting exactly a given list of accounts and no others?